### PR TITLE
Bugfix refactor blacklist and whitelist pattern matching

### PR DESCRIPTION
**Pull request by csaunders**
Please review @nickhoffman @xto

### CODE REVIEW

**Comment by xto** [comment]
:+1: Pretty sweet

**Comment by nickhoffman** [comment]
`ShopifyTheme.ignore_files` converts the `:ignore_files` config option [to regexes for us](https://github.com/Shopify/shopify_theme/blob/690f3ca448cb18b9de299fbc6a68dcf0e54d3cac/lib/shopify_theme.rb#L115-L117), but that doesn't happen for the `:whitelist_files` config option. It'd be good to do that for both or neither.

If we do it for both, the filters will convert the configured regexes to new regexes in their initializers. That isn't a problem (see below), but it'd be good to avoid it.

```ruby
r1 = /a/
# => /a/
r2 = Regexp.new(r1)
# => /a/
r1 == r2
# => true
```

File: spec/unit/cli_spec.rb [hunk 1]
```diff
@@ -41,6 +41,14 @@ def local_files
       assert_equal false, local_assets_list.include?('config.yml')
     end
 
+    it 'should only use the whitelist entries for determining which files to upload (bug #156)' do
+      @cli.local_files = %w(assets/application.css.liquid assets/application.js assets/image.png assets/bunny.jpg layout/index.liquid snippets/preview.liquid)
+      ShopifyTheme.config = {whitelist_files: %w(assets/application.css.liquid assets/application.js layout/ snippets/)}
+      local_assets_list = @cli.send(:local_assets_list)
```
Comment (nickhoffman) [inline spec/unit/cli_spec.rb]: Sneaky sneaky. Let's not call private methods.

File: spec/unit/file_filters_spec.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      end
+    end
+
+    it "initializing without a filter raises an error" do
+      assert_raises ArgumentError do
+        FileFilters.new
+      end
+    end
+
+    it "initializing with a single filter" do
```
Comment (nickhoffman) [inline spec/unit/file_filters_spec.rb]: This spec description refers to initializing, but the spec tests `#select` .

File: spec/unit/file_filters_spec.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      filters = FileFilters.new(IdentityFilter.new)
+      assert_equal [1, 2, 3, 4, 5], filters.select([1, 2, 3, 4, 5])
+    end
+
+    it "initializing with a list of filters" do
```
Comment (nickhoffman) [inline spec/unit/file_filters_spec.rb]: Same here re: spec description not matching the spec.

File: spec/unit/filters/blacklist_spec.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+          snippets/fancy.liquid
+      )
+
+      it "should return the entire list back if initialized with no patterns" do
+        blacklist = Blacklist.new
+        assert_equal TEST_PATHS, blacklist.select(TEST_PATHS)
+      end
+
+      it "should return everything except for the items that matched the pattern" do
+        blacklist = Blacklist.new %w(config/* settings.html assets/* templates/* snippets/*)
```
Comment (nickhoffman) [inline spec/unit/filters/blacklist_spec.rb]: :cop: parens plz

File: spec/unit/filters/whitelist_spec.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+          layout/thing1.html
+          assets/application.css.liquid
+          assets/application.js
+          templates/thing2.html
+          snippets/fancy.liquid
+        )
+        assert_equal expected, whitelist.select(TEST_PATHS)
+      end
+
+      it "should ignore the default whitelist if something was provided" do
```
Comment (nickhoffman) [inline spec/unit/filters/whitelist_spec.rb]: Nitpick: how about "if 1+ whitelists were provided"?

**Comment by nickhoffman** [comment]
Other than one comment about regexes and some test nitpicks, LGTM. :ship:

File: spec/unit/cli_spec.rb (hunk 1 above)
Comment (csaunders) [inline spec/unit/cli_spec.rb]: I'll just make it public. I need it in order to test a bunch of stuff :(

**Comment by csaunders** [comment]
@nickhoffman good catch on the `ignore_files`!

**Pull request merged**